### PR TITLE
Table export - fix for plain obj + add exportBlackList

### PR DESCRIPTION
**Pull request by LucasHocker**
### <strong>Pull Request</strong>

<hr>
Hi, thank you for contributing to Barista with this pull request (PR).

To ensure a fast process and merging of your PR please make sure it fulfills the
coding standards and contribution guidelines.

- A feature proposal has been provided, discussed and approved first.
- There is a meaningful description of the issue in GitHub (Screenshots are
  often helpful).
- If the PR introduces breaking-changes or deprecations it matches the following
  guidelines.
  - The commit message follows our commit guidelines.
  - Tests for the changes have been added (for bug fixes / features).
  - Docs have been added / updated (for bug fixes / features).

Please choose the type appropriate for the changes below: <br>

#### Type of PR

 Bugfix (non-breaking change which fixes an issue) 

#### Checklist

- [X] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by LucasHocker** [comment]
Based on testing with UI team, one fix and one feature are required:
- Table export feature did not work on a plain object (instead of DtDataSource) - Fixed
- Some columns do not make sense to export. - Added column blacklist to allow dev to exclude as desired.

Screen Shot 2022-01-26 at 9 36 09 PM

File: libs/barista-components/table/src/table.ts [hunk 1]
```diff
@@ -97,6 +98,7 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
   private _loading: boolean;
   private _destroy$ = new Subject<void>();
   private _showExportButton: boolean = false; //Revert to opt-in instead of opt-out per request
+  private _exportBlackList: string[] = [];
```
Comment (tomheller) [inline libs/barista-components/table/src/table.ts:101]: We are trying to use inclusive language here, please rename this
```suggestion
  private _exportBlockList: string[] = [];
```

File: libs/barista-components/table/src/table.ts [hunk 2]
```diff
@@ -366,19 +380,26 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
 
   /** @internal Exports the filtered source data from the dataSource. */
   // Note: this is different from the display text, see instead _exportDisplayData().
-  _exportFilteredData(): void {
-    const exportData = this._filteredData;
-    if (this.isEmptyDataSource || typeof exportData[0] != 'object') {
-      return;
+  _exportFilteredData(selectedData?: T[]): void {
+    //nothing to export
+    if (this.isEmptyDataSource) return;
```
Comment (tomheller) [inline libs/barista-components/table/src/table.ts:385]:
```suggestion
    if (this.isEmptyDataSource) {
      return;
    }
```

File: libs/barista-components/table/src/table.ts [hunk 3]
```diff
@@ -366,19 +380,26 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
...
-    const keys: string[] = Object.keys(exportData[0]);
-
-    if (!keys.length) {
-      return;
-    }
-
+    let keys: string[] = Object.keys(exportData[0]).filter(
+      (h: string) => !this.exportBlackList.includes(h),
+    );
+    if (!keys.length) return;
```
Comment (tomheller) [inline libs/barista-components/table/src/table.ts:402]:
```suggestion
    if (!keys.length) {
      return;
    }
```

File: libs/barista-components/table/src/table.ts [hunk 4]
```diff
@@ -366,19 +380,26 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
...
     }
 
     const csvObj = { csv: '' };
-    const keys: string[] = Object.keys(exportData[0]);
-
-    if (!keys.length) {
-      return;
-    }
-
+    let keys: string[] = Object.keys(exportData[0]).filter(
```
Comment (tomheller) [inline libs/barista-components/table/src/table.ts:399]:
```suggestion
    const keys: string[] = Object.keys(exportData[0]).filter(
```

File: libs/barista-components/table/src/table.ts [hunk 5]
```diff
@@ -413,39 +434,27 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
   }
 
   /** @internal Exports the filtered display data from the dataSource after being formatted by a displayAccessor. */
-  _exportDisplayData(exportData: T[] = this._filteredData): void {
-    if (this.isEmptyDataSource || typeof exportData[0] != 'object') {
-      return;
+  _exportDisplayData(selectedData?: T[]): void {
+    //nothing to export
+    if (this.isEmptyDataSource) return;
```
Comment (tomheller) [inline libs/barista-components/table/src/table.ts:439]:
```suggestion
    if (this.isEmptyDataSource) {
      return;
    }
```

File: package.json [hunk 6]
```diff
@@ -91,7 +91,7 @@
     "@angular/platform-server": "^12.2.8",
     "@angular/router": "^12.2.8",
     "@babel/cli": "^7.12.10",
-    "@commitlint/cli": "^15.0.0",
+    "@commitlint/cli": "^14.1.0",
```
Comment (tomheller) [inline package.json:94]: Please make sure you are not changing the package or package lock.json. I don't think that you need to change any dependencies for this fix.

**Review by tomheller — CHANGES_REQUESTED** [review]

**Comment by tomheller** [comment]
Please double check your history on this branch. this contains a lot of commits that should not be in there .

**Comment by LucasHocker** [comment]
Closing. Will resubmit after rebase and reset.

**Pull request closed without merging**